src/peft/tuners/lora/corda.py

```diff
@@ -0,0 +1,347 @@
+
+        # Calculate eigens
+        collect_eigens(model, lora_config, verbose)
+
+        # Crop CorDA eigens so that there's less to save
+        crop_corda_eigens(model, lora_config)
+
+        # Save cache to disk
+        if cache_file is not None:
+            cache: dict[str, Any] = {}
+            for name, module in target_modules(model, lora_config):
+                cache[f"{name}.corda_method"] = module.corda_method
+                cache[f"{name}.rank"] = module.rank
+                cache[f"{name}.eigens.S_WC"] = module.eigens.S_WC
+                cache[f"{name}.eigens.U_WC"] = module.eigens.U_WC
+                cache[f"{name}.eigens.V_WC"] = module.eigens.V_WC
+
+            os.makedirs(os.path.dirname(cache_file), exist_ok=True)
+            torch.save(cache, cache_file)
+
+
+@torch.no_grad()
+def calib_cov_distribution(
+    model: nn.Module,
+    config: LoraConfig,
+    run_model: Optional[Callable[[], None]],
+    hooked_model: Optional[nn.Module],
+    covariance_file: Optional[str],
+):
+    if covariance_file is not None and os.path.exists(covariance_file) and os.path.getsize(covariance_file) > 0:
+        all_covariance_matrix = torch.load(covariance_file, map_location=get_model_device(model))
+        for name, module in target_modules(model, config):
+            module.covariance_matrix = all_covariance_matrix[name]
+        return
+
+    if run_model is None:
+        raise ValueError("run_model must be specified when covariance file and cache file aren't built.")
+    if hooked_model is None:
+        hooked_model = model
+    hooked_model.eval()
+
+    def hook(module, input, output):
+        input = input[0].detach().squeeze(0).data  ## (context_length = 2048, dim)
+        if not config.corda_config.use_float16_for_covariance:
+            input = input.float()
+        input = input / torch.max(input).abs()
+
+        # check if input is valid
+        if torch.isnan(input).any() or torch.isinf(input).any():
+            raise ValueError("Invalid value found in input, please check your input data.")
+
+        # calculate covariance and check if it's valid
+        covariance = input.t().matmul(input)
+        if torch.isnan(covariance).any() or torch.isinf(covariance).any():
+            raise ValueError(
+                "Invalid value found in covariance. Please file an issue at https://github.com/huggingface/peft/issues."
+            )
+
+        # calculate mean and std
+        mean = input.mean(0)
+        std = input.std(0)
+
+        # add to module
+        module.sample_count += 1
+        module.covariance_matrix += covariance
+        module.mean += mean
+        module.std += std
+
+        # free memory
+        del covariance, input
+
+    for name, module in target_modules(hooked_model, config):
+        module.sample_count = 0
+        module.covariance_matrix = 0
+        module.mean = 0
+        module.std = 0
+        module.register_forward_hook(hook)
```

Comment:
Correct me if I'm wrong but these hooks are never unregistered, right? So the forward hook would be needlessly run during training time as well. In that case, it is probably worthwhile to save the handles returned by `register_forward_hook` and run `handle.remove()` afterwards.

Reply:
Great catch! I've added the logic to remove hooks, and corresponding test in `test_initialization.py` function `test_lora_corda_hook_unregister`.